File: Leetcode/1995/1995.cpp
class Solution {
public:
    int countQuadruplets(vector<int>& nums) {
        constexpr int MAXN = 305;
        vector<int> unmap[MAXN];
        memset(unmap, 0, sizeof unmap);
        int n = (int)nums.size();
        for (int i = 0; i < n; ++i) {
            unmap[nums[i]].emplace_back(i);
        }
        int ans = 0;
        int temp;
        for (int a = 0; a < n-2; ++a) {
            for (int b = a+1; b < n-1; ++b) {
                for (int c = b+1; c < n; ++c) {
                    int temp = nums[a]+nums[b]+nums[c];
                    if (!unmap[temp].empty()) {
                        if (c < unmap[temp].front()) {
                            ans += (int)unmap[temp].size();
                        }
                        else if (c > unmap[temp].back()) {
                            continue;
                        }
                        else {
                            auto iter = --upper_bound(unmap[temp].begin(),unmap[temp].end(),c);
                            ans += (int)unmap[temp].size() - 1 - (int)(iter - unmap[temp].begin());
                        }
                    }
                }
            }
        }
        return ans;
    }
};

File: Leetcode/1995/1995.go
func countQuadruplets(nums []int) int {
	ans := 0
	for i := 0; i < len(nums)-3; i++ {
		for j := i+1; j < len(nums)-2; j++ {
			for k := j+1; k < len(nums)-1; k++ {
				for n := k+1; n < len(nums); n++ {
					if nums[i]+nums[j]+nums[k] == nums[n] {
						ans++
					}
				}
			}
		}
	}
	return ans
}